lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs

```diff
@@ -136,6 +132,63 @@ import qualified Data.ByteString as BS
 import qualified Data.Text as T
 import qualified Data.Text.Encoding as T
 
+-- @AddressScheme@ correlates @key@s with @Addresses@.
+--
+-- In a @SeqState@ we need tell if an @Address@ was created using a given @key@.
+--
+-- A @key@ may corrspond to multiple @Address@es. We cannot* convert
+-- an address into a public key.
+--
+-- To be able to compare @key@s and addresses, we convert them both to
+-- fingerprints, which are hashes of the public keys.
+--
+-- @@
+--
+--            paymentAddress         delegationAddress
+--
+--    Address ◀────────────── Key ──────────────▶ Address
+--        ╲                                         ╱
+--          ╲                  │                  ╱
+--            ╲                │                ╱
+--              ╲              │              ╱
+--                ╲            │            ╱
+--                  ╲          │          ╱
+--                    ╲        │        ╱
+--                      ╲      │      ╱
+--                        ╲    │    ╱
+--                          ╲  │  ╱
+--                            ╲│╱
+--                            ▼▼▼
+--                        Fingerprint
+-- @@
+--
+-- *) Except in jormungandr
+--
+-- TODO: @addressToText@ and @addressFromText@ could /maybe/ be removed if we made
+-- Address an anbstract type. But something entirely differnt could be desired
+-- as well.
+data AddressScheme (key :: Depth -> * -> *) = AddressScheme
+    { addressFromKey
+        :: key 'AddressK XPub -> Address
+    , keyFingerprint
+        :: key 'AddressK XPub -> Either ErrMkKeyFingerprint KeyFingerprint
+    , addressFingerprint
+        :: Address -> Either ErrMkKeyFingerprint KeyFingerprint
+    , addressToText
+        :: Address -> Text
+    , addressFromText
+        :: Text -> Either TextDecodingError Address
+    , addressFromFingerprint
+        :: KeyFingerprint -> Address
+        -- TODO: Can we remove the need for this somehow?
+        -- I think we can, by parameterizing AddressPool over address.
+    } deriving Generic
+
+instance Show (AddressScheme k) where
+    show _ = "<AddressScheme>"
```

Comment:
We can probably do something a little nicer by requiring `Typeable k`.

Reply:
Hm... Good idea!